src/tests/roc_core/test_string_list.cpp

```diff
@@ -0,0 +1,148 @@
+ */
+
+#include <CppUTest/TestHarness.h>
+
+#include "roc_core/heap_allocator.h"
+#include "roc_core/helpers.h"
+#include "roc_core/string_list.h"
+
+namespace roc {
+namespace core {
+
+TEST_GROUP(string_list) {
+    HeapAllocator allocator;
+};
+
+TEST(string_list, empty) {
+    StringList sl(allocator);
+
+    LONGS_EQUAL(0, sl.size());
+    CHECK(sl.front() == NULL);
+}
+
+TEST(string_list, push_back) {
+    StringList sl(allocator);
+
+    LONGS_EQUAL(0, sl.size());
+    CHECK(sl.front() == NULL);
+
+    CHECK(sl.push_back("foo"));
+
+    LONGS_EQUAL(1, sl.size());
+    STRCMP_EQUAL("foo", sl.front());
+
+    CHECK(sl.push_back("bar"));
+
+    LONGS_EQUAL(2, sl.size());
+    STRCMP_EQUAL("foo", sl.front());
+}
+
+TEST(string_list, nextof) {
+    StringList sl(allocator);
+
+    CHECK(sl.push_back("foo"));
+    CHECK(sl.push_back("bar"));
+    CHECK(sl.push_back("baz"));
+
+    LONGS_EQUAL(3, sl.size());
+
+    const char* str;
+
+    str = sl.front();
+    STRCMP_EQUAL("foo", str);
+
+    str = sl.nextof(str);
+    STRCMP_EQUAL("bar", str);
+
+    str = sl.nextof(str);
+    STRCMP_EQUAL("baz", str);
+
+    str = sl.nextof(str);
+    CHECK(str == NULL);
+}
+
+TEST(string_list, uniq) {
+    StringList sl(allocator);
+
+    CHECK(sl.push_back("foo"));
+    CHECK(sl.push_back("bar"));
+    CHECK(sl.push_back("baz"));
+
+    LONGS_EQUAL(3, sl.size());
+
+    CHECK(sl.push_back_uniq("bar"));
+    CHECK(sl.push_back_uniq("qux"));
+
+    LONGS_EQUAL(4, sl.size());
+
+    const char* str;
+
+    str = sl.front();
+    STRCMP_EQUAL("foo", str);
+
+    str = sl.nextof(str);
+    STRCMP_EQUAL("bar", str);
+
+    str = sl.nextof(str);
+    STRCMP_EQUAL("baz", str);
+
+    str = sl.nextof(str);
+    STRCMP_EQUAL("qux", str);
+
+    str = sl.nextof(str);
+    CHECK(str == NULL);
+}
+
+TEST(string_list, clear) {
+    StringList sl(allocator);
+
+    CHECK(sl.push_back("foo"));
+
+    LONGS_EQUAL(1, sl.size());
+    STRCMP_EQUAL("foo", sl.front());
+
+    sl.clear();
+
+    LONGS_EQUAL(0, sl.size());
+    CHECK(sl.front() == NULL);
+
+    CHECK(sl.push_back("barbaz"));
+
+    LONGS_EQUAL(1, sl.size());
+    STRCMP_EQUAL("barbaz", sl.front());
+}
+
+TEST(string_list, exponential_growth) {
+    StringList sl(allocator);
+
+    const char* prev_front = sl.front();
+    int num_reallocs = 0;
+
+    int expected_reallocs[] = {
+        1, 1, 1, 1,                                    //
+        2, 2, 2, 2,                                    //
+        3, 3, 3, 3, 3, 3, 3, 3,                        //
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4 //
+    };
+
+    for (size_t n = 0; n < ROC_ARRAY_SIZE(expected_reallocs); n++) {
+        CHECK(sl.push_back("123456789abcdef,123456789abcdef"));
+
+        if (prev_front != sl.front()) {
+            num_reallocs++;
+            prev_front = sl.front();
+        }
+
+        LONGS_EQUAL(n + 1, sl.size());
+        LONGS_EQUAL(expected_reallocs[n], num_reallocs);
+    }
+}
+
```

Comment:
Lets add some tests for empty strings, something similar to:

```cpp
TEST(string_list, empty_strings) {
    StringList sl(allocator);

    CHECK(sl.push_back(""));
    CHECK(sl.push_back(""));
    CHECK(sl.push_back(""));

    LONGS_EQUAL(3, sl.size());

    const char* str = NULL;

    str = sl.front();
    STRCMP_EQUAL("", str);

    str = sl.nextof(str);
    STRCMP_EQUAL("", str);

    str = sl.nextof(str);
    STRCMP_EQUAL("", str);

    str = sl.nextof(str);
    CHECK(str == NULL);
}
```

?

Reply:
Probably we can also add the following test:

```cpp
TEST(string_list, push_back_twice) {
    StringList sl(allocator);

    const char* str = "foo";

    sl.push_back(str);
    sl.push_back(str);

    LONGS_EQUAL(2, sl.size());

    const char* next = sl.front();
    STRCMP_EQUAL("foo", next);

    next = sl.nextof(next);
    STRCMP_EQUAL("foo", next);

    next = sl.nextof(next);
    CHECK(next == NULL);
}
```

to show that we do not touch the input string (for demonstration purpose).

Reply:
> to show that we do not touch the input string (for demonstration purpose).

What do you mean by "touching"?

Reply:
I meant "touching" == "modifying".

Reply:
Done.